src/Microsoft.AspNetCore.TestHost/ResponseFeature.cs

```diff
@@ -36,33 +36,42 @@ public ResponseFeature()
 
         public void OnStarting(Func<object, Task> callback, object state)
         {
-            var prior = _responseStarting;
-            _responseStarting = () =>
+            var prior = _responseStartingAsync;
+            _responseStartingAsync = async () =>
             {
-                callback(state);
-                prior();
+                await callback(state);
+                await prior();
             };
         }
 
         public void OnCompleted(Func<object, Task> callback, object state)
         {
-            var prior = _responseCompleted;
-            _responseCompleted = () =>
+            var prior = _responseCompletedAsync;
+            _responseCompletedAsync = async () =>
             {
-                callback(state);
-                prior();
+                try
+                {
+                    await callback(state);
+                }
+                finally { }
```

Comment:
this try/finally doesn't do anything, your exception gets rethrown after the finally. `await prior();` should be in the finally block